File: pydynpd/common_functions.py
import math

import numpy as np
from scipy.sparse import csc_matrix


def lag(mat, lagged, N, lag_number, fill=np.NaN):
    height = int(mat.shape[0] / N)
    for i in range(N):
        start_row = i * height
        end_row = start_row + height
        mat_i = mat[start_row:end_row, :]
        lagged_i = lagged[start_row:end_row, :]

        lagged_i[0:lag_number, :] = fill
        lagged_i[lag_number:height, :] = mat_i[0:(height - lag_number), :]


def get_first_diff_table(ori_arr: np.ndarray, N: int):
    num_cols = ori_arr.shape[1]
    num_rows = ori_arr.shape[0]
    height = int(num_rows / N)

    lag_arr = np.zeros((num_rows, num_cols), dtype='float64')
    tbr_arr = np.zeros((num_rows, num_cols), dtype='float64')

    lag(ori_arr, lag_arr, N, 1)

    tbr_arr = ori_arr - lag_arr
    return tbr_arr


def get_fod_table(ori_arr: np.ndarray, N: int):
    num_rows = ori_arr.shape[0]
    height = int(num_rows / N)

    num_cols = ori_arr.shape[1]

    tbr = np.empty((num_rows, num_cols), dtype='float64')
    next_sum = np.empty((1, num_cols), dtype='float64')
    this_sum = np.empty((1, num_cols), dtype='float64')
    this_avg = np.empty((1, num_cols), dtype='float64')
    temp = np.empty((height, num_cols), dtype='float64')

    tbr[:] = np.NaN

    this_sum[:] = np.NaN

    for i in range(N):
        ori_i = ori_arr[i * height:(i * height + height), :]
        tbr_i = tbr[i * height:(i * height + height), :]
        temp.fill(np.NaN)
        next_sum.fill(np.NaN)
        next_count = 0
        for j in range(height - 2, -1, -1):

            if np.isnan(ori_i[range(j + 1, j + 2), :]).any(axis=1):
                this_count = next_count
                this_sum = next_sum
                temp[j, :] = temp[j + 1, :]
            else:
                this_count = next_count + 1

                this_sum = np.nansum(np.vstack([next_sum, ori_i[j + 1, :]]), axis=0)
                this_avg = this_sum * (1.0 / this_count)
                temp[j, :] = (ori_i[j, :] - this_avg) * math.sqrt(this_count / (this_count + 1))

            next_sum = this_sum
            next_count = this_count

        tbr_i[0, :] = np.NaN
        tbr_i[range(1, height), :] = temp[range(0, height - 1), :]

    return tbr


def sum_product(listOflist, n_rows):
    num_elements = len(listOflist)

    for i in range(n_rows):
        list_temp = []
        for j in range(num_elements):
            if type(listOflist[j]) == list:
                var_list = listOflist[j]
                list_temp.append(var_list[i])
            elif type(listOflist[j]) == np.ndarray:
                var_mat = listOflist[j]
                list_temp.append(var_mat)
            else:
                pass  # throw error
        temp = np.linalg.multi_dot(list_temp)
        if i == 0:
            tbr = temp
        else:
            tbr += temp

    return (tbr)


def Windmeijer(M2, _M2_XZ_W2, W2_inv, zs2, vcov_step1, Cx_list, z_list, residual1, N):
    D = np.empty((M2.shape[0], M2.shape[1]), dtype='float64')

    x_height = int(Cx_list.shape[0] / N)
    z_height = int(z_list.shape[0] / N)
    for j in range(0, Cx_list.shape[1]):

        for i in range(0, N):
            x = Cx_list[(i * x_height):(i * x_height + x_height), :]

            u = residual1[(i * x_height):(i * x_height + x_height), 0:1]
            z = z_list[(i * z_height):(i * z_height + z_height), :]

            xu = np.matmul(x[:, j:(j + 1)], u.transpose())
            temp = z @ (xu + xu.transpose()) @ z.transpose()
            # temp_zxuzt=z@ xu @ z.transpose()
            # temp=temp_zxuzt + temp_zxuzt.transpose()

            if i == 0:
                zxz = temp
            else:
                zxz += temp

        partial_dir = (-1.0 / N) * zxz

        Dj = np.linalg.multi_dot([_M2_XZ_W2, partial_dir, W2_inv, zs2])
        Dj = (-1) * Dj

        D[:, j:(j + 1)] = Dj

    # temp = np.multiply(N, M2) + np.multiply(N, np.matmul(D, M2)) + np.multiply(N, np.matmul(M2, D.transpose()))
    temp_D_M2 = D @ M2
    temp = np.multiply(N, M2) + np.multiply(N, temp_D_M2) + np.multiply(N, temp_D_M2.transpose())
    temp = temp + np.matmul(np.matmul(D, vcov_step1), D.transpose())
    #
    return (temp)


def make_sparse_list(arr_list):
    nrow = len(arr_list)
    new_list = []
    for i in range(nrow):
        arr = arr_list[i]
        new_arr = csc_matrix(arr)
        new_list.append(new_arr)

    return (new_list)

File: pydynpd/sandbox/pydynpd/pydynpd/cpp/Common_Functions.cpp
//
// Created by Tiger on 5/4/2022.
//

#include "Common_Functions.h"
#define EIGEN_INITIALIZE_MATRICES_BY_NAN

using namespace std;
using namespace Eigen;

vector<bool> row_has_nan(const RowMatrixXd &x)
{
    int num_rows=x.rows();
    vector<bool> tbr(num_rows);
    for (int i=0; i<num_rows; ++i){

        tbr[i]=!((x.row(i).array()==x.row(i).array()).all());
        
    }
   return tbr;
}

RowMatrixXd common_inv(Ref <RowMatrixXd> ori) {

    Eigen::FullPivLU <RowMatrixXd> Lu(ori);
	return ori.completeOrthogonalDecomposition().pseudoInverse();
//    if (Lu.isInvertible())
//        return ori.inverse();
//    else
//        return ori.completeOrthogonalDecomposition().pseudoInverse();


}

double standard_normalCDF(double x) // Phi(-∞, x) aka N(x)
{
    return std::erfc(-x / std::sqrt(2)) / 2;
}

/* The function checks if the array elements are consecutive
If elements are consecutive, then returns true, else returns
false */
bool areConsecutive(int* arr, int n)  //int arr[]
{
	//int n = sizeof(arr) / sizeof(arr[0]);
	if (n <  1)
		return false;

	int min = *min_element(arr, arr + n);

	int max = *max_element(arr, arr + n);

	if (max - min + 1 == n)
	{
		/* Create a temp array to hold visited flag of all elements.
		Note that, calloc is used here so that all values are initialized
		as false */
		bool *visited = (bool *)calloc(n, sizeof(bool));
		int i;
		for (i = 0; i < n; i++)
		{
			/* If we see an element again, then return false */
			if (visited[arr[i] - min] != false)
				return false;

			/* If visited first time, then mark the element as visited */
			visited[arr[i] - min] = true;
		}

		/* If all elements occur once, then return true */
		return true;
	}

	return false; // if (max - min  + 1 != n)
}

 
//void tokenize(std::string const &str, const char* delim,
//            std::vector<std::string> &out)
//{
//    char *token = strtok_s(const_cast<char*>(str.c_str()), delim);
//    while (token != nullptr)
//    {
//        out.push_back(std::string(token));
//        token = strtok_s(nullptr, delim);
//    }
//}

vector<string> splitString(string str, char splitter) {
	vector<string> result;
	string current = "";
	for (int i = 0; i < str.size(); i++) {
		if (str[i] == splitter) {
			if (current != "") {
				result.push_back(current);
				current = "";
			}
			continue;
		}
		current += str[i];
	}
	if (current.size() != 0)
		result.push_back(current);
	return result;
}
bool variable_exists(string var_name,  vector<string> &cols) {
	//for (string var_name : names)
	//	if (var_name == var_name)
	//		return true;

	//return false;


	// if (std::find(cols.begin(), cols.end(), var_name) == cols.end())
	// 	return false;
	// return true;
	if (getIndex(cols, var_name)>=0)
		return true;
	else
		return false;

}

int getIndex(vector<string> &v, string K)
{
    auto it = std::find(v.begin(), v.end(), K);
 
    // If element was found
    if (it != v.end())
    {
     

        int index = it - v.begin();
        return index;
    }
    else {
        // If the element is not
        // present in the vector
        return -1;
    }
}

void lag(Ref<RowMatrixXd> mat, Ref<RowMatrixXd> lagged, int N, int lag_number, double fill){
	int	height = mat.rows() / N;
	int width = mat.cols();
	//int start_row, end_row;

	for (int i=0;i<N; ++i){
		//start_row = i * height;
		//end_row = start_row + height;
		Ref<RowMatrixXd> mat_i = mat.block(i*height, 0, height, width);
		Ref<RowMatrixXd> lagged_i = lagged.block(i*height, 0, height, width);

		if (!isnan(fill))
			lagged_i.block(0,0, lag_number, width) = RowMatrixXd::Zero(lag_number, width);
		else
			lagged_i.block(0,0, lag_number, width) = RowMatrixXd(lag_number, width);

		lagged_i.block(lag_number,0, height-lag_number, width) = mat_i.block(0,0,height - lag_number, width);
	}
}

RowMatrixXd get_first_diff_table(Ref<RowMatrixXd> ori_arr, int N){
	int num_cols = ori_arr.cols();
	int num_rows = ori_arr.rows();
	//int height = num_rows / N;

	RowMatrixXd lag_arr (num_rows, num_cols);
	RowMatrixXd tbr_arr (num_rows, num_cols);

	lag(ori_arr, lag_arr, N, 1, NAN);

	tbr_arr = ori_arr - lag_arr;
	return tbr_arr;

}



RowMatrixXd	get_fod_table(Ref<RowMatrixXd> ori_arr, int N){
	int num_cols = ori_arr.cols();
	int num_rows = ori_arr.rows();
	int height = num_rows / N;

	RowMatrixXd tbr (num_rows, num_cols);
	RowMatrixXd next_sum (1, num_cols);
	RowMatrixXd this_sum (1, num_cols);
	RowMatrixXd this_avg (1, num_cols);
	

//	tbr = np.empty((num_rows, num_cols), dtype='float64')
//	next_sum = np.empty((1, num_cols), dtype='float64')
//	this_sum = np.empty((1, num_cols), dtype='float64')
//	this_avg = np.empty((1, num_cols), dtype='float64')
//	temp = np.empty((height, num_cols), dtype='float64')

//	tbr[:] = np.NaN

//	this_sum[:] = np.NaN
	int this_count;
	for (int i=0; i<N; ++i){
		Ref<RowMatrixXd> ori_i = ori_arr.block(i * height,0, height, num_cols);
		Ref<RowMatrixXd> tbr_i= tbr.block(i * height,0, height, num_cols);
		
		RowMatrixXd temp (height, num_cols);
		RowMatrixXd next_sum (1, num_cols);
		int next_count = 0;
		//for j in range(height - 2, -1, -1):
		
		for (int j=height-2; j>=0; --j){

			if (!(ori_i.row(j).array()==ori_i.row(j).array()).all()){
				this_count = next_count;
				this_sum = next_sum;
				temp.row(j) = temp.row(j + 1);
			}else{
				this_count = next_count + 1;
				for(int k=0; k<num_cols; ++k){
					double temp_next_sum, temp_ori;
					if (isnan(next_sum(0,k)))
						temp_next_sum=0;
					else
						temp_next_sum=next_sum(0,k);
					
					if (isnan(ori_i(j+1,k)))
						temp_ori=0;
					else
						temp_ori=ori_i(j+1,k);

					this_sum(0,k)=temp_next_sum+temp_ori;
				}
				
				this_avg = this_sum * (1.0 / this_count);
				temp.row(j) = (ori_i.row(j) - this_avg) * sqrt(this_count / (this_count + 1));
			}
			next_sum = this_sum;
			next_count = this_count;

			
		}



	
	tbr_i.row(0) = RowMatrixXd(1, num_cols);
	tbr_i.block(1,0, height-1, num_cols) = temp.block(0, 0, height - 1, num_cols);

	}


	return tbr;
}

void saveData(string fileName, RowMatrixXd  matrix)
{
    //https://eigen.tuxfamily.org/dox/structEigen_1_1IOFormat.html
    const static IOFormat CSVFormat(FullPrecision, DontAlignCols, ", ", "\n");
 
    ofstream file(fileName);
    if (file.is_open())
    {
        file << matrix.format(CSVFormat);
        file.close();
    }
}
